File: apps/syn-cli/src/syn_cli/commands/conversations.py
"""Conversation log commands — show JSONL log lines and session metadata."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.table import Table

from syn_cli._output import console, format_timestamp
from syn_cli.commands._api_helpers import api_get

app = typer.Typer(
    name="conversations",
    help="Inspect agent conversation logs",
    no_args_is_help=True,
)


@app.command("show")
def show(
    session_id: Annotated[str, typer.Argument(help="Session ID")],
    offset: Annotated[int, typer.Option(help="Line offset for pagination")] = 0,
    limit: Annotated[int, typer.Option(help="Max lines to show (max 500)")] = 100,
) -> None:
    """Show parsed conversation log lines for a session."""
    data = api_get(
        f"/conversations/{session_id}",
        params={"offset": offset, "limit": limit},
    )

    lines = data.get("lines", [])
    total = data.get("total_lines", 0)

    if not lines:
        console.print("[dim]No conversation lines found.[/dim]")
        return

    console.print(
        f"[bold]Conversation:[/bold] {session_id[:16]}  "
        f"[dim](lines {offset + 1}-{offset + len(lines)} of {total})[/dim]"
    )
    console.print()

    table = Table(show_edge=False, pad_edge=False)
    table.add_column("#", justify="right", style="dim", no_wrap=True)
    table.add_column("Type", style="cyan", no_wrap=True)
    table.add_column("Tool", style="dim", no_wrap=True)
    table.add_column("Preview")

    for line in lines:
        table.add_row(
            str(line.get("line_number", "")),
            line.get("event_type") or "—",
            line.get("tool_name") or "—",
            line.get("content_preview") or "—",
        )
    console.print(table)

    if total > offset + len(lines):
        next_offset = offset + limit
        console.print(f"[dim]More lines available. Use --offset {next_offset}.[/dim]")


def _format_field(value: object, fmt: str) -> str:
    """Format a metadata field value by format type."""
    if fmt == "ts":
        return format_timestamp(str(value))
    if fmt == "bytes":
        return f"{value:,} bytes"
    if fmt == ":,":
        return f"{value:,}"
    return str(value)


_METADATA_FIELDS: list[tuple[str, str, str]] = [
    ("model", "Model", ""),
    ("event_count", "Events", ":,"),
    ("total_input_tokens", "Input tokens", ":,"),
    ("total_output_tokens", "Output tokens", ":,"),
    ("started_at", "Started", "ts"),
    ("completed_at", "Completed", "ts"),
    ("size_bytes", "Log size", "bytes"),
]


@app.command("metadata")
def metadata(
    session_id: Annotated[str, typer.Argument(help="Session ID")],
) -> None:
    """Show metadata summary for a session's conversation."""
    m = api_get(f"/conversations/{session_id}/metadata")

    if m is None:
        console.print("[dim]No metadata recorded for this session.[/dim]")
        return

    console.print(f"[bold]Metadata:[/bold] {m.get('session_id', session_id)}")
    for key, label, fmt in _METADATA_FIELDS:
        value = m.get(key)
        if value is None:
            continue
        console.print(f"  {label + ':':17s}{_format_field(value, fmt)}")

    if m.get("tool_counts"):
        console.print("  Tool counts:")
        for tool, count in sorted(m["tool_counts"].items(), key=lambda x: -x[1]):
            console.print(f"    {tool}: {count}")

File: apps/syn-cli-node/src/commands/conversations.ts
/**
 * Conversation log commands — show, metadata.
 * Port of apps/syn-cli/src/syn_cli/commands/conversations.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { apiGet } from "../client/api.js";
import { print, printError, printDim } from "../output/console.js";
import { style, BOLD, CYAN, DIM } from "../output/ansi.js";
import { formatTimestamp } from "../output/format.js";
import { Table } from "../output/table.js";
import type { ConversationLogResponse, ConversationMetadataResponse } from "../generated/types.js";

const showCommand: CommandDef = {
  name: "show",
  description: "Show parsed conversation log lines for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  options: {
    offset: { type: "string", description: "Line offset for pagination", default: "0" },
    limit: { type: "string", description: "Max lines to show (max 500)", default: "100" },
  },
  handler: async (parsed: ParsedArgs) => {
    const sessionId = parsed.positionals[0];
    if (!sessionId) { printError("Missing session-id"); throw new CLIError("Missing argument", 1); }
    const offset = parseInt((parsed.values["offset"] as string) ?? "0", 10);
    const limit = parseInt((parsed.values["limit"] as string) ?? "100", 10);

    const data = await apiGet<ConversationLogResponse>(`/conversations/${sessionId}`, {
      params: { offset, limit },
    });

    if (data.lines.length === 0) { printDim("No conversation lines found."); return; }

    print(`${style("Conversation:", BOLD)} ${sessionId.slice(0, 16)}  ${style(`(lines ${offset + 1}-${offset + data.lines.length} of ${data.total_lines})`, DIM)}`);
    print("");

    const table = new Table();
    table.addColumn("#", { align: "right", style: DIM });
    table.addColumn("Type", { style: CYAN });
    table.addColumn("Tool", { style: DIM });
    table.addColumn("Preview");

    for (const line of data.lines) {
      table.addRow(
        String(line.line_number),
        line.event_type ?? "\u2014",
        line.tool_name ?? "\u2014",
        line.content_preview ?? "\u2014",
      );
    }
    table.print();

    if (data.total_lines > offset + data.lines.length) {
      printDim(`More lines available. Use --offset ${offset + limit}.`);
    }
  },
};

const metadataCommand: CommandDef = {
  name: "metadata",
  description: "Show metadata summary for a session's conversation",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const sessionId = parsed.positionals[0];
    if (!sessionId) { printError("Missing session-id"); throw new CLIError("Missing argument", 1); }

    const m = await apiGet<ConversationMetadataResponse>(`/conversations/${sessionId}/metadata`);

    print(`${style("Metadata:", BOLD)} ${m.session_id}`);

    if (m.model != null) print(`  Model:            ${m.model}`);
    if (m.event_count != null) print(`  Events:           ${m.event_count.toLocaleString()}`);
    if (m.total_input_tokens != null) print(`  Input tokens:     ${m.total_input_tokens.toLocaleString()}`);
    if (m.total_output_tokens != null) print(`  Output tokens:    ${m.total_output_tokens.toLocaleString()}`);
    if (m.started_at != null) print(`  Started:          ${formatTimestamp(m.started_at)}`);
    if (m.completed_at != null) print(`  Completed:        ${formatTimestamp(m.completed_at)}`);
    if (m.size_bytes != null) print(`  Log size:         ${m.size_bytes.toLocaleString()} bytes`);

    if (m.tool_counts) {
      print("  Tool counts:");
      const sorted = Object.entries(m.tool_counts).sort(([, a], [, b]) => b - a);
      for (const [tool, count] of sorted) {
        print(`    ${tool}: ${count}`);
      }
    }
  },
};

export const conversationsGroup = new CommandGroup("conversations", "Inspect agent conversation logs");
conversationsGroup.command(showCommand).command(metadataCommand);
